lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -153,12 +171,28 @@ void FIRRTLTypesLowering::runOnOperation() {
   while (!opsToRemove.empty())
     opsToRemove.pop_back_val()->erase();
 
-  // Remove args that have been lowered.
-  getOperation().eraseArguments(argsToRemove);
+  if (argsToRemove.empty())
+    return;
+
+  // Remove block args that have been lowered.
+  body->eraseArguments(argsToRemove);
   argsToRemove.clear();
 
+  // Copy over any attributes that weren't original argument attributes.
+  auto *argAttrBegin = originalArgAttrs.begin();
+  auto *argAttrEnd = originalArgAttrs.end();
+  std::sort(argAttrBegin, argAttrEnd);
```

Comment:
Can this use llvm::sort?

Reply:
Actually, I think this is already always sorted now, because DictionaryAttr always stors its attributes in sorted order (to canonicalize the dictionary).  I think you can even eliminate the originalArgAttrs copy and just use lowerbound over the elements in the dictionary attr itself.

Reply:
I wanted to take advantage of the fact that DictionaryAttr is sorted already, but I couldn't figure out how to track just the attributes we want to erase without copying them into some sort of array. The way that I'm doing this with `originalAttrs.getNamed(getArgAttrName(i))` builds that array in numerical order of arguments like `(arg0, arg1, arg2... arg10, arg11... arg20)`, but lower_bound expects them sorted like `(arg0... arg1, arg10, arg11... arg2, arg20)`. The FunctionSupport helper [getAllArgAttrs](https://github.com/llvm/llvm-project/blob/f7e675b3da1d7afd43af51c5a5e9140c87ef945d/mlir/include/mlir/IR/FunctionSupport.h#L314-L317) seems to build up the result in the same order, so I'm not sure if there is any way to track the original argument attributes and keep them sorted by name.

I was able to at least sink the loop that builds the array after the `if argsToRemove.empty()` check, and keep it sorted using the `DictionaryAttr::sortInPlace` helper (which has some optimizations and calls llvm::array_pod_sort in the general case) instead of `std::sort`.